rsfg/src/lexer.rs

```diff
@@ -6,175 +6,235 @@ static SPACE: &str = " \t";
+                        break;
+                    }
+                    lexer.rchars.pop();
+                }
+                let symbol_or_id = match text.as_ref() {
+                    "fn" => Fn,
+                    // These names clash, it sucks
+                    "int" => Token::Type(crate::Type::Int),
+                    _ => Identifier(text),
+                };
+                symbol_or_id
+            }
+            NextSymbolType::Space(c) => {
+                if lexer.tokens.last() == Some(&Newline) {
+                    if c == '\t' {
+                        Tab
+                    } else {
+                        // Figure out how many spaces we're using
+                        let mut count = 0;
+                        loop {
+                            match c {
+                                ' ' => {
+                                    count += 1;
+                                    if lexer.spaces_count != 0 && count == lexer.spaces_count {
+                                        break;
+                                    }
+                                }
+                                _ => break, // Ending on whitespace is fine
+                            };
+                        }
+                        if lexer.spaces_count == 0 && count != 0 {
+                            lexer.spaces_count = count;
+                        }
+                        if count == lexer.spaces_count {
+                            Tab
+                        } else {
+                            panic!("expected {} spaces, got {}", lexer.spaces_count, count);
+                        }
+                    }
+                } else {
+                    continue;
+                }
+            }
+            NextSymbolType::Digit(c) => {
+                let mut string = c.to_string();
+                loop {
+                    match lexer.rchars.last() {
+                        Some('0'...'9') => string.push(lexer.rchars.pop().unwrap()),
+                        Some('.') | Some('f') => panic!("floats not yet implemented"), // TODO
+                        _ => break,
+                    }
+                }
+                match string.parse() {
+                    Ok(number) => IntLit(number),
+                    Err(err) => panic!("{}", err),
+                }
+            }
+            NextSymbolType::Slash => {
+                match lexer.rchars.pop() {
+                    Some('/') => {
+                        // Comment
+                        while lexer.rchars.last() != None && lexer.rchars.last() != Some(&'\n') {
+                            lexer.rchars.pop();
+                        }
+                        continue;
+                    }
+                    _ => {
+                        // Division
+                        panic!("division operator not yet supported");
+                    }
+                }
+            }
+            NextSymbolType::Newline => Newline,
```

Comment:
i guess this here is the real payoff, not having to rchars.pop() every time, so nice